src/HideLooseAssetService.cpp

```diff
@@ -19,12 +22,69 @@ namespace BsaPacker
 			return false;
 		}
 
+		// Get QStringList of blacklisted extensions and remove the . so we can match with QFileInfo::suffix later
+		// This relies on the extension being in format .ext
+		QStringList blacklistExtensions = this->m_SettingsService->GetPluginSetting(SettingsService::SETTING_BLACKLISTED_FILES).toString().split(';');
+		for (auto& ext : blacklistExtensions) {
+			ext.remove(0, 1);
+		}
+
+		// loop through subdirectories in mod directory
 		const QString& absModDir = modDirectory.absolutePath();
 		for (const QString& subDir : modDirectory.entryList(QDir::Dirs | QDir::NoDotAndDotDot)) {
 			const QString& absPath = absModDir + '/' + subDir;
 			QDir dir(absPath);
-			if (!dir.dirName().endsWith(".mohidden")) {
-				dir.rename(absPath, absPath + ".mohidden");
+			
+			// directory is already hidden or empty so skip it
+			if (dir.dirName().endsWith(s_HiddenExt) || dir.isEmpty()) {
+				continue;
+			}
+
+			// get QStringList of file paths that should not be hidden and count total files
+			int fileCount = 0;
+			QStringList blacklistFilePaths;
+			QDirIterator iterator(absPath, QDir::Files, QDirIterator::Subdirectories);
+			while (iterator.hasNext()) {
+				QFileInfo fileInfo(iterator.nextFileInfo());
+				if (blacklistExtensions.contains(fileInfo.suffix(), Qt::CaseInsensitive)) {
+					blacklistFilePaths.append(fileInfo.absoluteFilePath());
+				}
+				++fileCount;
+			}
+
+			// directory has only blacklisted files so do nothing
+			if (fileCount == blacklistFilePaths.count())
+				continue;
+
+			// directory has no blacklisted files so hide it
+			if (blacklistFilePaths.isEmpty()) {
+				if (!dir.rename(absPath, absPath + s_HiddenExt)) {
+					qWarning() << "Failed to rename " << absPath << " to " << absPath + s_HiddenExt;
+				}
+				continue;
+			}
+
+			// directory has some blacklisted files so hide the directory then move blacklisted files out of it
+			if (!dir.rename(absPath, absPath + s_HiddenExt)) {
+				qWarning() << "Failed to rename " << absPath << " to " << absPath + s_HiddenExt;
+				continue;
+			}
+
+			const qsizetype absPathLength = absPath.length();
+			for (const auto& blacklistedPath : blacklistFilePaths) {
+				QFileInfo fileInfo(blacklistedPath);
+				QDir fileDir(fileInfo.absolutePath());
+
+				if (!fileDir.exists() && !fileDir.mkpath(fileDir.absolutePath())) {
+					qWarning() << "Failed to create path " << fileDir.absolutePath();
+					continue;
+				}
+
+				QString hiddenPath(blacklistedPath);
+				hiddenPath.insert(absPathLength, s_HiddenExt);
+				if (!fileDir.rename(hiddenPath, blacklistedPath)) {
+					qWarning() << "Failed to rename " << hiddenPath << " to " << blacklistedPath;
+				}
```

Comment:
To simplify all this code I would suggest:
- Iterating over all files and saving a list blacklisted files found
- Hiding all folders
- Move back the blacklisted files

You can group the list of blacklisted files by first folder if it makes things easier.

Reply:
Thank you for the feedback. I have simplified it to:
- Modify blacklistExtensions to use as a filter by prepending a *
- Hide all folders
- Iterate through and move back matching files